[FLINK-19582][network] Introduce sort-merge based blocking shuffle to Flink

wsry opened the pull request:
## What is the purpose of the change

Hash-based blocking shuffle and sort-merge based blocking shuffle are two main blocking shuffle implementations wildly adopted by existing distributed data processing frameworks. Hash-based implementation writes data sent to different reducer tasks into separate files concurrently while sort-merge based approach writes those data together into a single file and merges those small files into bigger ones. Compared to sort-merge based approach, hash-based approach has several weak points when it comes to running large scale batch jobs. By introducing the sort-merge based approach to Flink, we can improve Flink’s capability of running large scale batch jobs.


## Brief change log

  - Introduce SortBuffer and its implementation PartitionSortedBuffer for sort-merge based blocking shuffle
  - Introduce PartitionedFile and the corresponding writer/reader for sort-merge based blocking shuffle
  - Introduce sort-merge based result partition SortMergeResultPartition and the corresponding subpartition reader
  - Introduce new config options to enable sort-merge based blocking shuffle
  - Introduce shuffle data compression to sort-merge based blocking shuffle


## Verifying this change

Several new tests are added to verify this change, including PartitionSortedBufferTest, PartitionedFileWriteReadTest, SortMergeResultPartitionTest and BlockingShuffleITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

wsry commented:
I will do some minor update and this PR will be marked as ready for review by tomorrow.

wsry commented:
This PR is ready for review.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+
+	@Override
+	public void close() {
+		releaseCurrentSortBuffer();
```
`#close` could be called by canceller thread if I remembered correctly, then it might cause race condition here.
Maybe we can only release `fileWriter` as below to interrupt the task thread while writing data. For the other resource cleanup we can put them into `#releaseInternal()` before task exits.

wsry replied:
You are right, I ignored this.

zhijiangW replied:
If the variable `currentSortBuffer` might be touched concurrently by two threads (task main thread and canceler thread), it might still has the potential race condition if one thread reset it as `null` to cause misleading NPE for another thread.

The ideal condition is only allowing main thread access this variable and release it no matter with active or passive exit. Of course we can also resort to lock to avoid it if the logic is not easy to refactored.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+
+	private void writeCompressedBufferIfPossible(
+			Buffer buffer,
+			PartitionedFileWriter fileWriter,
```
The `fileWriter` is not necessary to provide as an argument since it can be got directly from `getPartitionedFileWriter()`, then it can simplify a bit the callers.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+		}
+
+		flushCurrentSortBuffer();
+		emit(record, targetSubpartition, dataType);
```
I think the below logic is not very easily to understand and also has some overheads in practice. 

```
if (!sortBuffer.hasRemaining()) {
	// the record can not be appended to the free sort buffer because it is too large
	releaseCurrentSortBuffer();
        writeLargeRecord(record, targetSubpartition, dataType);
	return;
}

flushCurrentSortBuffer();
emit(record, targetSubpartition, dataType);
```

Here we have to flush current buffer and call `emit` again to try out whether the unfilled sort buffer can hold the large record or not. During `sortBuffer.append` it would request all the segments from `LocalBufferPool` until exhausted before return `false` to decide flush large record.

I think of two options for improvement here.

- Make `SortMergeResultPartition` more light-weight component to only dispatch record/events, then the `SortBuffer` can decide itself when to append records in cache, when to flush them and when to flush large records directly internally. No need to expose many interfaces to interact with `SortMergeResultPartition`.

- `SortBuffer` provide another method to expose the cache threshold, it can be easily got from the internal `LocalBufferPool`. Then `SortMergeResultPartition` can decide whether to flush the large record directly to avoid unnecessary requesting all the segments.

wsry replied:
For the first option, the change is a little big and I would like to treat SortBuffer a purely in-memory data structure.
For the second option, the current size of buffer pool can be changed if new buffer pool created or old buffer pool is destroyed.

zhijiangW replied:
Thanks for the explanation. I might overlook the case of dynamic buffer pool for the second option.
I am not quite sure whether it is worth improving the overhead or simplifying the logic ATM. Maybe we can later decide it if necessary future.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+
+		if (!sortBuffer.hasRemaining()) {
+			// the record can not be appended to the free sort buffer because it is too large
+			releaseCurrentSortBuffer();
```
I guess this call is not really necessary. Because the current sort buffer can still be used in the followup small data after flushing the large record to file directly.

wsry replied:
The sort buffer can not be used, because the internal state has been broken after failing to write the large record. Maybe we can optimize the implementation of sort buffer to achieve that in the future.

zhijiangW replied:
Yeah, same thought as https://github.com/apache/flink/pull/13595#discussion_r514989760

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+	private SortBuffer currentSortBuffer;
+
+	/** File writer for this result partition. */
+	private PartitionedFileWriter fileWriter;
```
I guess this variable could be defined as final and setup in the constructor after removing `null` setting in below `finish()` method. Then we do not need to consider the condition of `fileWriter == null` is various usages.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortBuffer.java:
```diff
@@ -0,0 +1,96 @@
+	 * and returns {@link BufferWithChannel} which contains the copied data and the corresponding channel
+	 * index.
+	 */
+	BufferWithChannel copyData(MemorySegment target);
```
nit: rename as `copyIntoSegment` for better readable, otherwise I might be a bit confused whether `copyDataTo` or `copyDataFrom` in internal processes.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+	@Override
+	protected void releaseInternal() {
+		synchronized (lock) {
+			isFinished = true; // to fail writing faster
```
It seems not proper to misuse the variable of `isFinished` here. If we want to fail writing faster, we can make use of `isReleased` tag since it was also tagged true from `ResultPartition` before this execution. 

In other words, if the `releaseInternal()` is triggered during view consumption, the data writing should already be finished before. If it is triggered during writing data, then it should be executed by task thread, that means the writing process should already be exited before.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+			int subpartitionIndex = bufferWithChannel.getChannelIndex();
+
+			writeCompressedBufferIfPossible(buffer, fileWriter, subpartitionIndex);
+			updateStatistics(buffer, subpartitionIndex);
```
nit: `updateStatistics` could be covered inside `writeCompressedBufferIfPossible` since both `#writeLargeRecord` and `#flushCurrentSortBuffer`  have the same usages, then it can deduplicate the code a bit.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+
+	@Override
+	public void finish() throws IOException {
+		checkInProduceState();
```
nit: it is redundant here since already done in `super#finish`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+		flushCurrentSortBuffer();
+
+		synchronized (lock) {
+			checkState(!isReleased());
```
is this check really necessary? If so, i guess it should be placed onto `super#finish` and done firstly.

wsry replied:
It guarantees that the result partitioned file will not be produced if the result partition is already released.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+		releaseCurrentSortBuffer();
+
+		if (fileWriter != null) {
+			fileWriter.releaseQuietly();
```
It also has race condition for `fileWriter` as the similar case above. We can resolve this issue by defining it as final variable as commented https://github.com/apache/flink/pull/13595/files#r509983867

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+	}
+
+	private void flushCurrentSortBuffer() throws IOException {
+		if (currentSortBuffer == null || !currentSortBuffer.hasRemaining()) {
```
The condition of `currentSortBuffer == null` should not trigger the `releaseCurrentSortBuffer`. Maybe we can split them as below:

```
if (currentSortBuffer == null) {
     return;
}

if (currentSortBuffer.hasRemaining()) {
      .........
}
```

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java:
```diff
@@ -136,26 +144,41 @@ public ResultPartition create(
+					bufferCompressor,
+					bufferPoolFactory);
+			} else {
+				final BoundedBlockingResultPartition blockingPartition = new BoundedBlockingResultPartition(
```
directly write as `partition = xxx` ?

wsry replied:
We need the type of BoundedBlockingResultPartition when calling initializeBoundedBlockingPartitions.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+			buffer = bufferCompressor.compressToIntermediateBuffer(buffer);
+		}
+		fileWriter.writeBuffer(buffer, targetSubpartition);
+		buffer.recycleBuffer();
```
it would be safe to recycle the buffer in `finally` region in case of any exceptions in above processes. If so we might not reuse the same variable `buffer` as the return value of `bufferCompressor.compressToIntermediateBuffer`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+
+	@Override
+	public void broadcastEvent(AbstractEvent event, boolean isPriorityEvent) throws IOException {
+		Buffer buffer = EventSerializer.toBuffer(event, isPriorityEvent);
```
Make `buffer = EventSerializer.toBuffer(event, isPriorityEvent);` inside `try` region to avoid buffer leak once exception.

wsry replied:
If it throws an exception, the buffer must be null. we have nothing to release.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:
```diff
@@ -0,0 +1,360 @@
+			int numTargetKeyGroups,
+			int networkBufferSize,
+			ResultPartitionManager partitionManager,
+			FileChannelManager channelManager,
```
nit: might be better to pass `Path` instead of big component `channelManager` as constructor argument.

zhijiangW reviewed (commented):
I only finished the review of `SortMergeResultPartition` class, will continue on the left parts later.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBuffer.java:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	private int getLow32BitsFromLongAsInteger(long value) {
+		return (int) (value & 0xffff);
```
This only gets the low 16 bits, not the low 32 bits.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBuffer.java:
```diff
@@ -0,0 +1,390 @@
+		checkState(!isReleased, "Sort buffer is already released.");
+
+		int totalBytes = source.remaining();
+		if (totalBytes == 0) {
```
Is empty record possible ? I think if it is possible we should also write index for it, otherwise we should throw exceptions for illegal state, but it seems not reasonable to ignore it silently.

wsry replied:
Empty record is impossible here, I will add a check here.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:
```diff
@@ -173,6 +173,28 @@
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
+	public static final ConfigOption<Integer> NETWORK_MAX_BUFFERS_PER_SORT_MERGE_PARTITION =
+		key("taskmanager.network.sort-merge-blocking-shuffle.max-buffers-per-partition")
```
Can we make the keys a bit shorter? For example use `taskmanager.network.sort-shuffle.max-buffers`.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:
```diff
@@ -173,6 +173,28 @@
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
+	public static final ConfigOption<Integer> NETWORK_SORT_MERGE_SHUFFLE_MIN_PARALLELISM =
+		key("taskmanager.network.sort-merge-blocking-shuffle.min-parallelism")
```
Similar, maybe use `taskmanager.network.sort-shuffle.min-parallelism` here.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:
```diff
@@ -173,6 +173,28 @@
+	public static final ConfigOption<Integer> NETWORK_MAX_BUFFERS_PER_SORT_MERGE_PARTITION =
+		key("taskmanager.network.sort-merge-blocking-shuffle.max-buffers-per-partition")
+			.defaultValue(2048)
+			.withDescription("Maximum number of network buffers can be used per sort-merge blocking result partition. " +
```
Please add the `intType()` here to make it runtime type safe (and not have the deprecation warnings).

StephanEwen commented on flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:
```diff
@@ -173,6 +173,28 @@
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
+	public static final ConfigOption<Integer> NETWORK_SORT_MERGE_SHUFFLE_MIN_PARALLELISM =
+		key("taskmanager.network.sort-merge-blocking-shuffle.min-parallelism")
+			.defaultValue(Integer.MAX_VALUE)
```
Please add the `intType()` here to make it runtime type safe (and not have the deprecation warnings).

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java:
```diff
@@ -215,10 +246,19 @@ private static void releasePartitionsQuietly(ResultSubpartition[] partitions, in
 				numberOfSubpartitions * networkBuffersPerChannel + floatingNetworkBuffersPerGate : Integer.MAX_VALUE;
+			int numRequiredBuffers = numberOfSubpartitions + 1;
+
+			if (isSortMergeBlockingShuffle(type, numberOfSubpartitions)) {
```
Maybe rewrite this to not compute it twice in the sort shuffle mode (once for non-sort shuffle, then for sort shuffle).

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java:
```diff
@@ -166,6 +189,14 @@ else if (type == ResultPartitionType.BLOCKING || type == ResultPartitionType.BLO
 		return partition;
 	}
 
+	private boolean isBlockingShuffle(ResultPartitionType type) {
```
You don't need this, you can just say `!type.isPipelined()`.

StephanEwen reviewed (commented):
Thanks a lot, the code looks generally good to me.

I think the current configuration works also as a feature flag, which is nice. If I read it correctly, sort-shuffle will start from a parallelism of INT_MAX, which means it is never used unless users explicitly configure it to be used. So we do not need any additional feature flag.

What confused my were the minimum and maximum buffer requirements.

  - There is a maximum number of buffer (`taskmanager.network.sort-merge-blocking-shuffle.max-buffers-per-partition`). Why do we need that? What would happen if there were more buffers?

  - We also still have a minimum of `numSubpartitions + 1` buffers. Is this a requirement for the write-phase, or only for the read phase? As I understand it, a part of the motivation of this sort-nased shuffle is to decouple the number of needed memory buffers from the parallelism. Could we drop this minimum?

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,211 @@
+	private void flushIndexBuffer() throws IOException {
+		if (indexBuffer.position() > 0) {
+			indexBuffer.flip();
+			indexFileChannel.write(indexBuffer);
```
I think here we still need to iterates till the whole buffer is write (writeFully). The dataFileChannel has also done this.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBuffer.java:
```diff
@@ -0,0 +1,390 @@
+	 * Size of an index entry: 4 bytes for record length, 4 bytes for data type and 8 bytes
+	 * for pointer to next entry.
+	 */
+	private static final int INDEX_ENTRY_SIZE = 4 + 4 + 8;
```
As a whole, I think the index entry size could be reduced into 12 bytes by 4 bytes record length + 4 bytes for segment index + 3 bytes for segment offset + 1 bytes for data type. This would help increase the number of records in each region, especially when the record size is smaller. But we might also leave it as a future optimization (if we think it is ok)

wsry replied:
Good point, I can further optimize it in the future PRs.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFile.java:
```diff
@@ -0,0 +1,139 @@
+	 * Returns the index entry offset of the target region and subpartition in the index file. Both region index
+	 * and subpartition index start from 0.
+	 */
+	private long getIndexEntryOffset(int region, int subpartition) {
```
I think it does not make much difference here to return long since we always first cast it to integer when using the result, we may change the return type to int and use `MathUtils.checkedCastDown` in return statement.

wsry replied:
We need long when reading from file

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java:
```diff
@@ -0,0 +1,193 @@
+	private final Queue<Buffer> buffersRead = new ArrayDeque<>();
+
+	/** Target subpartition to read. */
+	private final int subpartitionIndex;
```
This variable only used in constructor

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,211 @@
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		currentSubpartition = Math.max(currentSubpartition, targetSubpartition);
```
I am a bit confused why we need to maintain this variable for above `checkArgument(targetSubpartition >= currentSubpartition...`. I do not see the necessary from it.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,211 @@
+
+	private void writeRegionIndex() throws IOException {
+		if (Arrays.stream(subpartitionBytes).sum() > 0) {
+			for (int subpartition = 0; subpartition < numSubpartitions; ++subpartition) {
```
I guess the current way assumes the data redistributed in almost all the subpartitions for every flush call. If in the data skew scenario like only 10 of 1000 subpartitions have actual data for some flush, then the index buffer will be written many empty holes to waste space and execution time. 

I am curious that is it by design to have to write all the subpartitions in the same sequence since I have not touched the detail logics on view reader side? Or it is not very performance sensitive for easy implementation ATM?

wsry replied:
At reading side, we use (subpartition index) × (index entry size) locate an index entry for the given subpartition in a data region. And you are right, it's not very performance sensitive.

zhijiangW replied:
Ok, let's further drive into it if really necessary future.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,211 @@
+	 */
+	public PartitionedFile finish() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
```
The `dataFileChannel` and `indexFileChannel` are created via data output driven. I am wondering that does it exist the case of empty output? Maybe the operator filters all the input data without any output, then finished finally.

If so, we should remove above check to avoid unnecessary failure.

wsry replied:
We always have an EndOfPartition event

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFile.java:
```diff
@@ -0,0 +1,139 @@
+	private final Path indexFilePath;
+
+	/** Used to accelerate index data access. */
+	private final ByteBuffer indexDataCache;
```
nit: better to add annotation `@Nullable`, also for the respective constructor argument.

zhijiangW commented:
@wsry , I am not quite sure whether we ever had the case of empty data for some subpartitions in any form of tests. 
If not, we should verify such scenarios to guarantee no potential bugs in this new supplement partition.  As based on my previous experience, it is easy to trigger some corner case in logics if some subpartitions without any data until job finishes.

gaoyunhaii commented on flink-tests/src/test/java/org/apache/flink/test/runtime/BlockingShuffleITCase.java:
```diff
@@ -0,0 +1,100 @@
+
+		@Override
+		public void invoke(String value) throws Exception {
+			assert value.equals(RECORD);
```
might change to `assertEquals` ?

wsry commented:
> @wsry , I am not quite sure whether we ever had the case of empty data for some subpartitions in any form of tests.
> If not, we should verify such scenarios to guarantee no potential bugs in this new supplement partition. As based on my previous experience, it is easy to trigger some corner case in logics if some subpartitions without any data until job finishes.

I will add one for blocking shuffle.

wsry commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBufferTest.java:
```diff
@@ -0,0 +1,291 @@
+
+		// fill the sort buffer with randomly generated data
+		int totalBytesWritten = 0;
+		SortBuffer sortBuffer = createSortBuffer(bufferPoolSize, bufferSize, numSubpartitions);
```
I think there is no need to do that, because these resources will be cleared after test finishes.

wsry commented:
> Thanks a lot, the code looks generally good to me.
> 
> I think the current configuration works also as a feature flag, which is nice. If I read it correctly, sort-shuffle will start from a parallelism of INT_MAX, which means it is never used unless users explicitly configure it to be used. So we do not need any additional feature flag.
> 
> What confused my were the minimum and maximum buffer requirements.
> 
> * There is a maximum number of buffer (`taskmanager.network.sort-merge-blocking-shuffle.max-buffers-per-partition`). Why do we need that? What would happen if there were more buffers?
> * We also still have a minimum of `numSubpartitions + 1` buffers. Is this a requirement for the write-phase, or only for the read phase? As I understand it, a part of the motivation of this sort-nased shuffle is to decouple the number of needed memory buffers from the parallelism. Could we drop this minimum?

@StephanEwen 
1. As suggested by @zhijiangW Zhijiang in the discussion thread, I replace `taskmanager.network.sort-shuffle.min-parallelism` with a feature flag `taskmanager.network.sort-shuffle.enabled` for ease of use.
2. You are right, the motivation of `taskmanager.network.sort-shuffle.max-buffers` is used to decouple the number of needed memory buffers from the parallelism. So it is the upper bound of required memory, if `numSubpartitions + 1` is greater than this config value, this config value is used. The config option is currently used by write-phase, after we implement partition request restriction or buffer management in FLINK-16641, this option can also be reused by read phase. Then we can totally decouple the number of needed memory buffers from the parallelism for both write-phase and read-phase.

wsry commented:
@StephanEwen @zhijiangW @gaoyunhaii Thanks for the review and comments, I have updated the PR according to the review comments.

StephanEwen commented:
@wsry 

I am confused a bit about the use of max buffers. That value is the upper limit of buffers that will be assigned to the shuffle from the global pool. It is only used for streaming pipelined connections, to limit the amount of in-flight buffers (so checkpoints don't take too long). For batch it should not be used, because there is no harm in using as much memory as possible.

The min buffers is actually hat decouples the memory use from the parallelism. If the min-buffers is related to the number of subpartitions, then we still have the problem that shuffles fail on large parallelism (parallelism is higher than available memory buffers).

So, in conclusion, I think we should not have a max value (because it does not help in decoupling from parallelism) and also need to decouple the min value from the parallelism.

I also liked the idea of having the `taskmanager.network.sort-shuffle.min-parallelism` as a flag. That way low parallelisms (< 50) could use the hash shuffle and larger parallelisms could use the sort shuffle.

wsry commented:
@StephanEwen 
`taskmanager.network.sort-shuffle.max-buffers` is only used to control the max required buffers (min size) of the buffer pool now. I have update the code and the calculation logic is as follows:
```
int maxNumberOfMemorySegments = type.isBounded() ? numberOfSubpartitions * networkBuffersPerChannel + floatingNetworkBuffersPerGate : Integer.MAX_VALUE;
int numRequiredBuffers = !type.isPipelined() && sortMergeBlockingShuffleEnabled ? Math.min(numberOfSubpartitions + 1, maxBuffersPerSortMergePartition) : numberOfSubpartitions + 1;

return bufferPoolFactory.createBufferPool(numRequiredBuffers, maxNumberOfMemorySegments, numberOfSubpartitions, maxBuffersPerChannel);
```

I agree that `taskmanager.network.sort-shuffle.min-parallelism` is more flexible than a boolean flag and I will update the PR and use `taskmanager.network.sort-shuffle.min-parallelism`.

StephanEwen commented:
The updated logic looks good. But the name `maxBuffersPerSortMergePartition` seems confusing. I think we should rename this to `minBuffersPerSortMergePartition` and also rename the config key to `taskmanager.network.sort-shuffle.min-buffers`.

What is a good default for the minimum number of buffers? Maybe something like 512?

wsry commented:
> The updated logic looks good. But the name `maxBuffersPerSortMergePartition` seems confusing. I think we should rename this to `minBuffersPerSortMergePartition` and also rename the config key to `taskmanager.network.sort-shuffle.min-buffers`.
> 
> What is a good default for the minimum number of buffers? Maybe something like 512?

@StephanEwen These are good suggestions. `taskmanager.network.sort-shuffle.min-buffers` is easier to understand from the perspective of user and set the default value to 512 is also reasonable. I have updated the PR accordingly.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,230 @@
+		}
+
+		try {
+			dataFileChannel.close();
```
Looks like there is a copy/paste error here, and this should be `indexFileChannel`.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,230 @@
+		try {
+			dataFileChannel.close();
+		} catch (Throwable throwable) {
+			exception = throwable;
```
This way, you'll lose the first exception. I would either keep the first exception or use the `ExceptionUtils.firstOrSuppressed()` pattern here.

StephanEwen commented:
@wsry I have two more questions on the minimum buffers logic and confi

(1)

I previously thought that this is to ensure that the sort buffers are not too small. But I saw now that they can actually be smaller, because the minimum is `Math.min(minBuffersPerSortMergePartition , numSubpartitions + 1)`. So if we have a low parallelism, then the sort shuffle might run with something like 5 buffers (assuming a user turns it on for all shuffles). Is that on purpose?

Does it make sense to just use the `minBuffersPerSortMergePartition` value and not lower it below that?

BTW: I think I now understand why you called it "max buffers" before - because it is the maximum to which the minimum requirement can grow.

(2)

The current value for `taskmanager.network.sort-shuffle.min-buffers` may be high for mini clusters. At least if we end up doing what is proposed in (1) and not lower it for lower parallelism.

The current min is 512 buffers (16 MB) and the mini cluster has only 64 MB network memory by default. So as soon as there are four shuffles, the mini cluster will break, which is pretty soon.
We should probably set a different min on the mini cluster, like 64 buffers or so.

wsry commented:
> @wsry I have two more questions on the minimum buffers logic and confi
> 
> (1)
> 
> I previously thought that this is to ensure that the sort buffers are not too small. But I saw now that they can actually be smaller, because the minimum is `Math.min(minBuffersPerSortMergePartition , numSubpartitions + 1)`. So if we have a low parallelism, then the sort shuffle might run with something like 5 buffers (assuming a user turns it on for all shuffles). Is that on purpose?
> 
> Does it make sense to just use the `minBuffersPerSortMergePartition` value and not lower it below that?
> 
> BTW: I think I now understand why you called it "max buffers" before - because it is the maximum to which the minimum requirement can grow.
> 
> (2)
> 
> The current value for `taskmanager.network.sort-shuffle.min-buffers` may be high for mini clusters. At least if we end up doing what is proposed in (1) and not lower it for lower parallelism.
> 
> The current min is 512 buffers (16 MB) and the mini cluster has only 64 MB network memory by default. So as soon as there are four shuffles, the mini cluster will break, which is pretty soon.
> We should probably set a different min on the mini cluster, like 64 buffers or so. This does not require a new config key, it you can just set this directly in the mini cluster configuration.

Using `taskmanager.network.sort-shuffle.min-buffers` as the min required number of buffers directly is also a reasonable choice. Previously, I was thinking that users can switch to sort-merge shuffle directly without changing the config value (may need to increase network buffers for small parallelism jobs). However, to achieve a better performance, the user may have to tune these config values, if so, leaving the config values unchanged seems meaningless.

I will update the PR soon as suggested.

StephanEwen commented:
Thanks for explaining, @wsry 

Maybe we can find a good combination of "min-parallelism" and "min-buffers" to make the switch to sort-based shuffle work well for users. Because then the very low parallelism cases would use the partition-based shuffle and the higher-parallelism cases the sort-based shuffle.

wsry commented:
@StephanEwen 
It is a really good point. If the "min-parallelism" is larger than the "min-buffers", the network buffers will be always satisfied.
Then what do you think if we give a smaller default value for "min-buffers", for example 64. If so, we do not need to change the default mini-cluster configuration, besides, if there are more available buffers in the global network buffer pool, we can also use them, because the max number of buffer can be used by the local buffer pool is Integer.MAX.

StephanEwen commented:
@wsry Your suggestion sounds good to me!

wsry commented:
@StephanEwen I have updated the PR accordingly.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java:
```diff
@@ -0,0 +1,242 @@
+		}
+
+		if (exception != null) {
+			ExceptionUtils.rethrow(exception);
```
Nit: You can add `throws Exception` to the method signature (compatible with `AutoClosable`) and then you don't need to wrap it as a RuntimeException.

wsry replied:
I use ExceptionUtils.rethrow for it can handle both Error and Exception.

StephanEwen replied:
In this case it does a redundant wrapping into RuntimeException. If you add the exception signature you avoid that.
Exception wrapping should be avoided unless you add actual information in the new exception. Otherwise it just complicates the stack traces for debugging.

StephanEwen commented:
Looks good form my side!

Let's wait for @zhijiangW and @gaoyunhaii and CI to give their ok as well.

We can start looking at documentation in the meantime. I think we need a quick description of this and how to configure it in the memory configuration section: 
  - https://ci.apache.org/projects/flink/flink-docs-master  ==> Deployment & Operations ==> Memory Configuration
  - Or add it to the tuning page: https://ci.apache.org/projects/flink/flink-docs-master/ops/memory/mem_tuning.html

gaoyunhaii commented:
Very thanks for @wsry  opening the PR and @StephanEwen @zhijiangW for the reviewing! The PR also looks good to me in general from my side and thus +1 for merging. A small issue is that if available, I still tend to that we add some tests covering the case some subpartitions are empty in the SortBuffer or in one region, as @zhijiangW pointed out before, since we have some logic related to skip the empty sub-partitions. Of course if not available we may also add the tests in the separate PR.

wsry commented:
Thanks for @gaoyunhaii for pointing out the test issue, I misunderstand @zhijiangW 's suggestion before and and add an IT case totally without data. I will supplement some UT case to cover the scenario.

zhijiangW commented:
Thanks for the updates @wsry ! 

I think most of my previous comments were addressed except for https://github.com/apache/flink/pull/13595#discussion_r514988635. And as @gaoyunhaii also mentioned above, it is better to supplement some tests for covering empty subpartition case.

StephanEwen commented:
Looks like the concerns are all addressed. @zhijiangW and @gaoyunhaii any remaining open issues from your side?

gaoyunhaii commented:
There are no issues from my side and +1 for merging!

wsry commented:
> Thanks for the updates @wsry !
> 
> I think most of my previous comments were addressed except for [#13595 (comment)](https://github.com/apache/flink/pull/13595#discussion_r514988635). And as @gaoyunhaii also mentioned above, it is better to supplement some tests for covering empty subpartition case.

@zhijiangW 
I have addressed the race condition by adding a lock. Besides, tests with empty subpartitions are also added, including PartitionSortedBufferTest#testWriteReadWithEmptyChannel and PartitionedFileWriteReadTest#testWriteAndReadWithEmptySubpartition

wsry commented:
> Looks good form my side!
> 
> Let's wait for @zhijiangW and @gaoyunhaii and CI to give their ok as well.
> 
> We can start looking at documentation in the meantime. I think we need a quick description of this and how to configure it in the memory configuration section:
> 
> * https://ci.apache.org/projects/flink/flink-docs-master  ==> Deployment & Operations ==> Memory Configuration
> * Or add it to the tuning page: https://ci.apache.org/projects/flink/flink-docs-master/ops/memory/mem_tuning.html

@StephanEwen 
I added the document to memory tuning page in commit fe68ecb10c60d28ecff3ea3030a0b6b3555e5e29 ([FLINK-19602][network] Introduce new config options to enable sort-merge based blocking shuffle) and the azure pipeline is also green now.

zhijiangW commented:
I still have a bit concern for the handling of `currentSortBuffer` in the race condition case. If the canceler thread resets it as `null` via `#close`, then the main task thread might be involved in `NullPointerException` and exits. My previous assume is to exit main thread via closed file channel like destroyed buffer pool in stream task mode. The NPE might confuse users to find the root cause and more seem like a bug. 

If possible, we can avoid reset `currentSortBuffer` as `null` while releasing, then it would simplify the related logics a bit.

StephanEwen commented:
@zhijiangW  As soon as a task is cancelled, exception in the main thread are no longer reported. That is exactly because it is very hard to guarantee in each individual component that no exceptions occur once something was cancelled/closed. At least that is how it was for a long time, I hope the mailbox didn't break this behavior.

So the exception coming out in the main thread should be ignored here, in my understanding.

wsry commented:
> I still have a bit concern for the handling of `currentSortBuffer` in the race condition case. If the canceler thread resets it as `null` via `#close`, then the main task thread might be involved in `NullPointerException` and exits. My previous assume is to exit main thread via closed file channel like destroyed buffer pool in stream task mode. The NPE might confuse users to find the root cause and more seem like a bug.
> 
> If possible, we can avoid reset `currentSortBuffer` as `null` while releasing, then it would simplify the related logics a bit.

@zhijiangW I have updated the PR to avoid the potential issue.

wsry commented:
> @zhijiangW As soon as a task is cancelled, exception in the main thread are no longer reported. That is exactly because it is very hard to guarantee in each individual component that no exceptions occur once something was cancelled/closed. At least that is how it was for a long time, I hope the mailbox didn't break this behavior.
> 
> So the exception coming out in the main thread should be ignored here, in my understanding.

@StephanEwen Thanks for explaining.

zhijiangW commented:
Thanks for the explanation @StephanEwen and the updates @wsry !

My previous understanding was that the thrown exception from main thread would still be shown in TM log, but will not reported to JM in the case of cancel. If so users might still see NPE from TM log to get confused sometimes. Our users always inquired for the BufferPool destroyed exception from TM log in the case of cancel before, not only concentrate on JM log. Or my knowledge is out of date/misunderstood. 

It might be a tiny issue enlarged by me. Anyway, @wsry already avoided it now. I am also +1 for merging!

StephanEwen commented:
Nice, thanks everyone for the work 👍 

I'll merge this as soon as the CI passes.

@zhijiangW Let us take it a QA issue for 1.12 that we don't show / log exceptions after cancellation.
Maybe there are some utility classes that log exceptions without knowing the context. This is a good example why logging exceptions locally is tricky, because one might not know the context there, whether it is an expected exception or so.
